### DISCUSSION TITLE
Single select using or_ with and_ expressions

### DISCUSSION THREAD

**Discussion by natej** [sqlalchemy/sqlalchemy#6848] [Usage Questions] [answered]
Hi,

I'm currently generating and running multiple SELECTs and would like to construct a single SELECT that's the equivalent of the code below:

```python
orders = []
for widget in widgets:
    stmt = select(Order).where(expression.and_(Order.widgetid == widget.id,
                                               Order.last == widget.last))
    orders.extend(session.execute(stmt).scalars().all())
```

For example, how could I take:
```python
ands = [expression.and_(Order.widgetid == widget.id,
                        Order.last == widget.last) for widget in widgets]
```
and use it with `expression.or_` and ultimately `where()`?

I'm using SA 1.4.22 with PostgreSQL. I'm passing `future=True` to `Session()` and `create_engine()`. I'm trying to move from `session.query()` to `select()` in 2.0-style and get ORM entities directly.

Thanks for your help.

**Comment by CaselIT** [comment, marked as answer]
Hi,

Sure, that would work.
You may also use [`tuple_`](https://docs.sqlalchemy.org/en/14/core/sqlelement.html?highlight=tuple#sqlalchemy.sql.expression.tuple_) with an in_. Something like
```py
select(Order).where(tuple_(Order.widgetid, Order.last).in_([(widget.id, widget.last) for widget in widgets]))
```

**Reply by natej** [reply]
Hey, thanks so much! That works great.

In case someone else finds this, I'm going to paste below what I ended up using. It's the same as yours except for `,in_` should be `.in_`. `in_()` is an operator available on a column:

```python
stmt = select(Order).where(
    expression.tuple_(Order.widgetid, Order.last).in_(
        [(widget.id, widget.last) for widget in widgets]
    )
)
result = session.execute(stmt).scalars().all()
```

Thanks again!

**Reply by CaselIT** [reply]
Oh sorry that was a typo

**Reply by natej** [reply]
No worries, I knew it was. :) I just mentioned it in case someone finds this in the future and it helps them.
